contrib/platform/src/com/sun/jna/platform/win32/COM/util/ProxyObject.java

```diff
@@ -0,0 +1,627 @@
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import com.sun.jna.Pointer;
+import com.sun.jna.WString;
+import com.sun.jna.platform.win32.Guid;
+import com.sun.jna.platform.win32.Guid.IID;
+import com.sun.jna.platform.win32.Guid.REFIID;
+import com.sun.jna.platform.win32.Kernel32;
+import com.sun.jna.platform.win32.Kernel32Util;
+import com.sun.jna.platform.win32.OaIdl;
+import com.sun.jna.platform.win32.OaIdl.DISPID;
+import com.sun.jna.platform.win32.OaIdl.DISPIDByReference;
+import com.sun.jna.platform.win32.OaIdl.EXCEPINFO;
+import com.sun.jna.platform.win32.OleAuto;
+import com.sun.jna.platform.win32.OleAuto.DISPPARAMS;
+import com.sun.jna.platform.win32.Variant;
+import com.sun.jna.platform.win32.Variant.VARIANT;
+import com.sun.jna.platform.win32.Variant.VariantArg;
+import com.sun.jna.platform.win32.WTypes;
+import com.sun.jna.platform.win32.WinDef;
+import com.sun.jna.platform.win32.WinDef.DWORDByReference;
+import com.sun.jna.platform.win32.WinDef.LCID;
+import com.sun.jna.platform.win32.WinDef.UINT;
+import com.sun.jna.platform.win32.WinNT;
+import com.sun.jna.platform.win32.WinNT.HRESULT;
+import com.sun.jna.platform.win32.COM.COMException;
+import com.sun.jna.platform.win32.COM.COMUtils;
+import com.sun.jna.platform.win32.COM.ConnectionPoint;
+import com.sun.jna.platform.win32.COM.ConnectionPointContainer;
+import com.sun.jna.platform.win32.COM.Dispatch;
+import com.sun.jna.platform.win32.COM.IDispatch;
+import com.sun.jna.platform.win32.COM.IDispatchCallback;
+import com.sun.jna.platform.win32.COM.util.annotation.ComInterface;
+import com.sun.jna.platform.win32.COM.util.annotation.ComMethod;
+import com.sun.jna.platform.win32.COM.util.annotation.ComProperty;
+import com.sun.jna.ptr.IntByReference;
+import com.sun.jna.ptr.PointerByReference;
+
+/**
+ * Ole32.INSTANCE.CoInitialize must be called on the current thread before using
+ * this object
+ */
+public class ProxyObject implements InvocationHandler, com.sun.jna.platform.win32.COM.util.IDispatch,
+		IRawDispatchHandle {
+
+	public ProxyObject(Class<?> theInterface, IDispatch rawDispatch, Factory factory) {
+		this.unknownId = -1;
+		this.rawDispatch = rawDispatch;
+		this.comThread = factory.getComThread();
+		this.theInterface = theInterface;
+		this.factory = factory;
+		//make sure dispatch object knows we have a reference to it
+		// (for debug it is usefult to be able to see how many refs are present
+		int n = this.rawDispatch.AddRef();
+		this.getUnknownId(); //pre cache it
+		factory.register(this);
+	}
+	
+	//cached value of the IUnknown interface pointer
+	// Rules of COM state that querying for the IUnknown interface must return an identical pointer value
+	long unknownId;
+	long getUnknownId() {
+		if (-1 == this.unknownId) {
+			try {
+	
+				final PointerByReference ppvObject = new PointerByReference();
+	
+//				HRESULT hr = this.comThread.execute(new Callable<HRESULT>() {

```

Comment:
Erm, sorry. I think there could be a few places where I haven't had time to do a really good code clean. Feel free to point out odd things like this.


Using JNA COM via the COM.utils classes should not need you to worry about threads. All COM calls are executed on their own thread, but this is hidden under the ProxyObject and other API, ending in calls executed within the 'ComThread' class.

If I remember correctly, I had to comment out this specific use of the comThread because it caused a non terminating recursive loop of calls. As it turns out, the call to QueryInterface, doesn't have to be performed on THE COM thread, so I just commented out the thread wrapping code in case I needed to put it back. However all is working find to I have now removed the commented out code.

And done a few other minor code cleaning bits. In particular the comment at the top of ProxyObject was out-of-date.